### PR TITLE
[RFC] Replace vim_strbyte() with standart strchr()

### PR DESCRIPTION
**Pull request by ghost**
Replace vim_strbyte() with @aktau wrapper that uses the standart strchr() #1476 . Fix the duplicate code in vim_strchr() because the issues are closely related #1474 . Add tests.

### CODE REVIEW

**Comment by justinmk** [comment]
Did you check the existing PRs?

**Comment by ivegotasthma** [comment]
I did, the PRs that I found were #2502, #2848 and #1489. They all seem to be abandoned. I see that in #2502 the submitter did do some benchmarks. @aktau also showed a benchmarking script in #1476 which I used it to test it. I am not sure if I should write a benchmark for this.

**Comment by justinmk** [comment]
See my comment: https://github.com/neovim/neovim/pull/2502#issuecomment-98423110

**Comment by ivegotasthma** [comment]
From #1476 
> Wow, nice results. vim_strchr/vim_strbyte/cstrchr are on the hot path of bt_regexec_both() so this could speed up syntax highlighting, among other things

I've never used a profiler before but after viewing `strings.c` in it to test the highlighting and searching through a lot of text for a couple of minutes but I didn't see a call to `bt_regex_both()`.

The largest amount of calls to `vim_strchr()` seem to be made from `cstrchr()` followed by `skip_range()` and `find_vat_ht()`. I searched with both regexes and normal words.

When `cstrchr()` is called this leads to a call to `vim_strchr()` in almost 25% of the cases out of which 50% are calls to `utfc_ptr2len()`.

How can I profile this further? What should I do with this benchmark, where should I record it?
I can't find much information on the web about `bt_regexec_both()` and since there are no calls to it I am unsure how to invoke it. I checked the source code but searching with regexes didn't call it. Any further advice?

**Comment by justinmk** [comment]
> I can't find much information on the web about bt_regexec_both() and since there are no calls to it 

It depends on the regex pattern used in the search. For example, it is called by `bt_regexec_multi` which means a multi-line pattern would trigger it. Another thing to check out is `regmatch` which ends up looping `cstrchr` in a few cases.

> How can I profile this further? What should I do with this benchmark, where should I record it?

Take 10 samples of `time nvim -u NONE -c [some search command repeated multiple times]' on 3 sample files (one with ASCII text, one with utf8 text, one with long lines). Do the test with the old regex engine, the new "NFA", and "autoselect" using `\%#=0/1/2` (see `:help NFA`)). Compare the results before/after. Try a multiline pattern like `/foo\_.bar` and `/foo\nbar`, as well as non-multiline pattern.

If the results seem promising, then we'll go from there.

**Comment by ivegotasthma** [comment]
For the text files I used 5 paragraphs of lorem ipsum in 3 files. I used iconv on one of to convert it to ascii, the other two were utf8, one of which with pretty long lines.

The multiline search command I used was `./nvim -u NONE -c "set re={1,2}" -c "/\w\+[\n\s]\w\+" -c "bn" -c "bn" -c "bn" -c "q!" ascii utf8 long`

In these cases the most calls were made from `copy_option_part` ~ 90% (17k out of 19k calls), which in turn was mostly called by `shada_removable()`.

The singleline search command is `./nvim -u NONE -c "set re={1, 2}" -c "/\w\+\s\w\+" -c "bn" -c "bn" -c "bn" -c "qa!" ascii utf8 long`

The non regex search command `./nvim -u NONE -c "set re=0" -c "/lor" -c "bn" -c "bn" -c "bn" -c "qa!" ascii utf8 long`

I also tested with a non-regex search and my results were pretty much the same as with a regex on both engines. Am I doing something wrong?

I think I got a lot of calls to `cstrchr` because I was moving around the highlighted results and higlighting multiple times, now that I am getting cleaner input I see that `vim_strchr` is on the hot path of `copy_option_part`.

What should I be looking for, because I am feeling lost. 

Is the input I am giving you okay, when do/would the results be promising?

**Comment by justinmk** [comment]
Can you get times (average of 10 runs) before and after the change?

**Comment by ivegotasthma** [comment]
I made a small script to time the execution of the different commands with different versions. The results are at the end of the [script](https://gist.github.com/ivegotasthma/07154d4859545aa0ff25)

The new version is quite faster.

File: src/nvim/strings.c [hunk 1]
```diff
@@ -420,14 +416,19 @@ char_u *vim_strchr(const char_u *string, int c)
 char_u *vim_strbyte(const char_u *string, int c)
   FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
 {
-  const char_u *p = string;
+  if (c == 0) {
+    return NULL; /* Vim prefers returning NULL instead of the end of string */
```
Comment (ZyX-I) [inline src/nvim/strings.c]: return NULL;  // …

File: src/nvim/strings.c [hunk 2]
```diff
@@ -420,14 +416,19 @@ char_u *vim_strchr(const char_u *string, int c)
...
+  }
 
-  while (*p != NUL) {
-    if (*p == c)
-      return (char_u *) p;
-    ++p;
+  if (c > (unsigned char) -1) {
+    /* vim_strbyte() never casts c to int, which means that if c >= 255, it
```
Comment (ZyX-I) [inline src/nvim/strings.c]: `//` comment

File: src/nvim/strings.c [hunk 3]
```diff
@@ -420,14 +416,19 @@ char_u *vim_strchr(const char_u *string, int c)
...
-  const char_u *p = string;
+  if (c == 0) {
+    return NULL; /* Vim prefers returning NULL instead of the end of string */
+  }
 
-  while (*p != NUL) {
-    if (*p == c)
-      return (char_u *) p;
-    ++p;
+  if (c > (unsigned char) -1) {
```
Comment (ZyX-I) [inline src/nvim/strings.c]: 255. I do not see why you are using `(unsigned char) -1`.

File: src/nvim/strings.c [hunk 4]
```diff
@@ -420,14 +416,19 @@ char_u *vim_strchr(const char_u *string, int c)
...
+    /* vim_strbyte() never casts c to int, which means that if c >= 255, it
+     * simply never matches and will always return NULL, but not before
+     * running through the entire string first. We emulate this behaviour by
+     * just returning NULL immediately. */
+    return NULL;
   }
-  return NULL;
+
+  return (unsigned char*) strchr((char*) string, c);
```
Comment (ZyX-I) [inline src/nvim/strings.c]: `(char_u *)`.

**Comment by justinmk** [comment]
The results do seem promising. 
Did you compile both case in Release or RelWithDebInfo?

Nice effort!

**Comment by ivegotasthma** [comment]
That's from a Debug build. The results from the RelWithDebInfo are [here](https://gist.github.com/ivegotasthma/d41bea897f49c94c2ce2). It seems that there isn't almost any difference when optimisations are applied, I am guessing that I need to use more complex regexes to notice a difference.

**Comment by justinmk** [comment]
No, it's not surprising that they are similar--just wanted to be sure.

**Comment by aktau** [comment]
Thanks for jumping on this and doing a proper job @ivegotasthma :). I'm surprised that any differences where noticeable with 5 paragraphs of lorem ipsum. That should be way too short to notice a difference in even the most inefficiently written `strchr`-like.

Debug builds _could_ make a large difference versus Release builds with the old vim_strbyte, since it's open-coded the compiler could do all sorts of inlining and possibly even light vectorizing. For the new version, it shouldn't make a realy difference since the libc `strchr` is the real meat and it's always optimized. I remember benchmarking with my script from #1476 but can't remember whether it made an actual difference.

**Comment by ivegotasthma** [comment]
@aktau Thanks. I will test it on very large blocks of text. 

[Here](https://gist.github.com/ivegotasthma/c45ca6ef378e43db39c3) is the result, it's pretty clear now when I used around 1.3k 80col lines or 315 ~4-5 multiline lines. There's a pretty consistent 100ms improvement across regex searches and 300 ms improvement for non regex search.

@justinmk Is there something more I need to add to this PR to improve it?

**Comment by justinmk** [comment]
I and/or someone else need to verify the results then we can move forward. The second result set suggests that the first result set was dominated by some other operation (maybe shada?). Though even 1s seems high for such small files--what kind of machine is this running on?

**Comment by ivegotasthma** [comment]
I am using Arch Linux as a distro.
CPU: Intel(R) Core(TM) i5-2430M CPU @ 2.40GHz
Kernel:  4.2.5-1 x86_64 GNU/Linux
CC: gcc 5.2.0

**Comment by ticki** [comment]
Is this PR dead?

**Comment by ivegotasthma** [comment]
I think it was finished, last time I checked, it was awaiting an approval.

**Comment by ticki** [comment]
Yeah, that sounds very strange... Perhaps, @ivegotasthma mistakenly confused the unit for another one?

**Comment by justinmk** [comment]
Closing in favor of https://github.com/neovim/neovim/pull/6460. Thanks for looking into it.

**Pull request closed without merging**